Meta/Lagom/Tools/CodeGenerators/LibWeb/GenerateWebGLRenderingContext.cpp

```diff
@@ -713,17 +713,36 @@ class @class_name@ {
             }
             function_impl_generator.set("number_of_vector_elements", number_of_vector_elements);
             function_impl_generator.append(R"~~~(
+    @cpp_element_type@ const* data = nullptr;
+    size_t count = 0;
     if (v.has<Vector<@cpp_element_type@>>()) {
-        auto& data = v.get<Vector<@cpp_element_type@>>();
-        glUniform@number_of_vector_elements@@gl_postfix@v(location->handle(), data.size() / @number_of_vector_elements@, data.data());
-        return;
+        auto& vector = v.get<Vector<@cpp_element_type@>>();
+        data = vector.data();
+        count = vector.size();
+    } else if (v.has<GC::Root<WebIDL::BufferSource>>()) {
+        auto& typed_array_base = static_cast<JS::TypedArrayBase&>(*v.get<GC::Root<WebIDL::BufferSource>>()->raw_object());
+        auto& typed_array = verify_cast<JS::@typed_array_type@>(typed_array_base);
+        data = typed_array.data().data();
+        count = typed_array.array_length().length();
+    } else {
+        VERIFY_NOT_REACHED();
     }
+)~~~");
 
-    auto& typed_array_base = static_cast<JS::TypedArrayBase&>(*v.get<GC::Root<WebIDL::BufferSource>>()->raw_object());
-    auto& typed_array = verify_cast<JS::@typed_array_type@>(typed_array_base);
-    @cpp_element_type@ const* data = typed_array.data().data();
-    auto count = typed_array.array_length().length() / @number_of_vector_elements@;
-    glUniform@number_of_vector_elements@@gl_postfix@v(location->handle(), count, data);
+            // FIXME: If srcOffset + srcLength is longer than data.length, generate INVALID_VALUE.
```

Comment:
we should have the ability to generate INVALID_VALUE by calling `set_error` on the Context. I added a virtual for it that will get forwarded to the real WebGL/Context object with the first commit for WebGL2 with the IDL files.